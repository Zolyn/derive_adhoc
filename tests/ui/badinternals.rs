// Tests that cause internal macros to generate their "incompatible versions"
// advice.  Actually generating this advice in a more realistic scenario,
// rather than just directly invoking the macros, is quite hard.  So this
// test is not particularly realistic.  But it does exercise the error
// generation code paths.

use derive_adhoc::{define_derive_adhoc, Adhoc};

#[derive(Adhoc)]
struct Driver;

define_derive_adhoc! {
    Template =
}

derive_adhoc_driver_Driver!{ GARBAGE -> DRIVER }
derive_adhoc_template_Template!{ GARBAGE -> TEMPLATE }
derive_adhoc::derive_adhoc_expand!{ GARBAGE -> INNARDS }

// OpCompatVersions 1.0, ie our own, and the earliest released
derive_adhoc::derive_adhoc_expand!{
    { pub struct StructName {} }
    [1 0]
    { }
    { pub struct OptOk0; }
    { crate; [] Template; }
}

// OpCompatVersions 1.200, some imaginary future compatible one
derive_adhoc::derive_adhoc_expand!{
    { pub struct StructName {} }
    [1 200]
    { }
    { pub struct OptOk200; }
    { crate; [] Template; }
}

// OpCompatVersions 200.1, some imaginary future incompatible one
derive_adhoc::derive_adhoc_expand!{
    { pub struct StructName {} }
    [200 0]
    { }
    { OUGHT NOT TO BE EXPANDED; }
    { crate; [] Template; }
}

fn main() {
    let _ = OptOk0;
    let _ = OptOk200;
}
