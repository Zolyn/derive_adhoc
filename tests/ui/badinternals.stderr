error: wrong input to derive-adhoc driver inner macro derive_adhoc_driver_Driver; might be due to incompatible derive-adhoc versions(s)
  --> ui/badinternals.rs:9:10
   |
9  | #[derive(Adhoc)]
   |          ^^^^^
...
16 | derive_adhoc_driver_Driver!{ GARBAGE -> DRIVER }
   | ------------------------------------------------ in this macro invocation
   |
   = note: this error originates in the macro `derive_adhoc_driver_Driver` (in Nightly builds, run with -Z macro-backtrace for more info)

error: wrong input to derive-adhoc template inner macro derive_adhoc_template_Template; might be due to incompatible derive-adhoc versions(s)
  --> ui/badinternals.rs:12:1
   |
12 | / define_derive_adhoc! {
13 | |     Template =
14 | | }
   | |_^
...
17 |   derive_adhoc_template_Template!{ GARBAGE -> TEMPLATE }
   |   ------------------------------------------------------ in this macro invocation
   |
   = note: this error originates in the macro `derive_adhoc_template_Template` (in Nightly builds, run with -Z macro-backtrace for more info)

error: bad input to derive_adhoc_expand inner template expansion proc macro; might be due to incompatible derive-adhoc versions(s)
  --> ui/badinternals.rs:18:1
   |
18 | derive_adhoc::derive_adhoc_expand!{ GARBAGE -> INNARDS }
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `derive_adhoc::derive_adhoc_expand` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected curly braces
  --> ui/badinternals.rs:18:37
   |
18 | derive_adhoc::derive_adhoc_expand!{ GARBAGE -> INNARDS }
   |                                     ^^^^^^^
